src/groupByKey.ts

```diff
@@ -2,6 +2,17 @@ type GroupsMap<T> = {
   [key: string]: T[];
 };
 
-export function groupByKey(items, key) {
-  // write code here;
+export function groupByKey<T>(items: T[], key: keyof T): GroupsMap<T> {
+  const groupedByKey: GroupsMap<T> = {};
+  const allKeys = items
+    .map((item) => item[key])
+    .filter((item, index, arr) => arr.indexOf(item) === index);
```

Comment:
The current approach to obtaining unique keys is inefficient because it involves two iterations over the 'items' array - one for mapping and another for filtering. This can be optimized by using a Set to collect unique keys in a single iteration.